README
Site-Engine
===========

Site::Engine - Ugly tiny CGI-based web-framework with templates, sessions and
               databases.

This is simple CGI-based web-framework. It's syntax looks like Dancer
(http://perldancer.org). Use this only if your hosting has plain perl
and Apache with mod_cgi/mod_rewrite and doesn't provide support for
modern PSGI-based WF like Dancer or Mojolicious. Also, it will be easy
to migrate your code later to Dancer if needed.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  CGI.pm

EXAMPLE

    #!/usr/bin/perl
    #
    # app.pl
    #
    use strict;
    use warnings;
    use Site::Engine;

    # Configuration of site
    my %config = (
        "templates" => "/path/to/templates/dir",
    );

    # Routes
    get qr{/} => sub {
        template 'index', {
            data => "hello world!"
        }, { layout => 'main' }
    };

    start_site \%config;

DEPLOYMENT

Sample apache 1/2 config

<VirtualHost *>
    ServerName example.com
    DocumentRoot /srv/example.com
    <Directory /srv/example.com>
        Options ExecCGI +SymLinksIfOwnerMatch
        AddHandler cgi-script .pl
        RewriteEngine On
        RewriteBase /
        RewriteCond %{REQUEST_URI} !\.(js|css|png|jpg|gif)$
        RewriteCond %{REQUEST_URI} !\/app(\.pl)*\/
        RewriteRule ^(.*)$ /app.pl/$1 [L]
        ErrorDocument 404 /app.pl/404
    </Directory>
</VirtualHost>

COPYRIGHT AND LICENCE

Copyright (C) 2011 by crux <thecrux@gmail.com>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.12.3 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Site::Engine.

0.01  Tue May 31 22:25:45 2011
	- original version; created by h2xs 1.23 with options
		-AX -b 5.8.8 -n Site::Engine -v 0.01



lib/Site/Engine/Session.pm
package Site::Engine::Session;
use strict;
use warnings;
use Fcntl;

our $VERSION = '0.01';

my $config;

# Public

sub init ($) {
    $config = shift;
    die "failed to setup session dir"
      if ( !exists $config->{"session_dir"}
        || !-d $config->{"session_dir"} );
    $config->{session_expires} = $config->{session_expires} || 60 * 30;
}

sub destroy_session ($$) {
    my $file = _session_file(@_);
    if ( defined $file ) {
        unlink $file;
    }
}

sub session {
    my ( $id, $addr, $field, $value ) = @_;

    my $file = _session_file( $id, $addr );

    if ( defined $file ) {
        my $time = time;
        my $dt = $time - ( stat($file) )[9];
        if ( $dt < $config->{session_expires} ) {
            open my $fh, "<", $file or die $!;
            my %s;
            while (<$fh>) {
                chomp();
                my ( $var, @value ) = split /=/, $_;
                $s{$var} = join "=", @value;
            }
            close $fh;
            if ( defined $value && defined $field && $field ne "id" ) {
                $s{$field} = $value;
                open my $fh, ">", $file or die $!;
                foreach my $key ( keys %s ) {
                    print $fh $key . "=" . $s{$key} . "\n";
                }
                close $fh;
            }
            elsif ( defined $field ) {
                $value = ( $field eq "id" ) ? $id : $s{$field};
            }
            else {
                $s{id} = $id;
                $value = \%s;
            }
            utime $time, $time, $file;
            return ( $id, $value );
        }
        else {

            # delete old session file
            unlink $file;
        }
    }
    $id = &_generate_session( $addr, $field, $value );
    return ($id);
}

# Private

sub _session_file ($$) {
    my ( $id, $addr ) = @_;

    # is addr sane
    return if ( $addr !~ /^[\d\.]+$/ );

    my $file = $config->{"session_dir"} . "/" . $addr . "-";
    if (   defined $id
        && length $id == 16
        && $id !~ /\W/
        && -f $file . $id )
    {
        return $file . $id;
    }
    return;
}

sub _generate_session ($$) {
    my ( $addr, $field, $value ) = @_;

    # is addr sane
    return if ( $addr !~ /^[\d\.]+$/ );

    my $id;
    foreach my $i ( 0 .. 9 ) {
        $id = join '',
          ( "_", 0 .. 9, "a" .. "z", "A" .. "Z" )
          [ map { int( rand 63 ) } ( 1 .. 16 ) ];
        if (
            sysopen my $fh,
            $config->{"session_dir"} . "/" . $addr . "-" . $id,
            O_CREAT | O_EXCL | O_WRONLY
          )
        {
            print $fh $field . "=" . $value . "\n"
              if ( defined $value && defined $field && $field ne "id" );
            close($fh);
            last;
        }
        else {
            $id = undef;
        }
    }
    $id;
}

1;
__END__

=pod

=head1 NAME

Site::Engine::Sesion - Session handling

=head1 METHODS

=head2 init( \%config )

init module with configuration

=over

=item "session_dir"     - existed and writable directory for session files

=item "session_expires" - lifetime in seconds of session

=back
        
=head2 session( $id, $addr, $key, $value ) - get/set values for current session.

=over

=item $id    - id (16 random symbols \w) of current session or undef (to create new session)

=item $addr  - ip-address of client (REMOTE_ADDR)

=item $key   - get/set some attribute of current session (if undefined - return ref to hash with all values)

=item $value - set attribute $key with specified $value (if undefined - function is getter)

=back

=head2 destroy_session( $id, $addr ) - destroy session with $id and $addr

=over

=item $id    - id (16 random symbols \w) of current session or undef (to create new session)

=item $addr  - ip-address of client (REMOTE_ADDR)

=back

=head1 DESCRIPTION

This module provide simple session support based on a plain files (one per
session) with key=value in each line. Name of session file is $addr-$id

=head1 COPYRIGHT AND LICENSE

Copyright 2011 by crux E<lt>thecrux@gmail.comE<gt>

This module is free software and is published under the same terms as Perl itself.

=cut


lib/Site/Engine/Database.pm
package Site::Engine::Database;
use strict;
use warnings;
use DBI;
use Exporter qw( import );
our @EXPORT  = qw( dbh );
our $VERSION = '0.01';

my $config;
my $dbh;
my $dbi = {
    sqlite => sub {
        my $db_conf = shift;
        DBI->connect( "dbi:SQLite:dbname=" . $db_conf->{db_file},
            "", "", { RaiseError => 1, sqlite_unicode => 1 } );
    },
    mysql => sub {
        my $db_conf = shift;
        DBI->connect(
            "DBI:mysql:database=" 
              . $db_conf->{db}
              . (
                ( exists $db_conf->{db_host} ) ? ";host=" . $db_conf->{db_host}
                : ""
              )
              . (
                ( exists $db_conf->{db_port} ) ? ";port=" . $db_conf->{db_port}
                : ""
              ),
            $db_conf->{db_user},
            $db_conf->{db_pass},
            { RaiseError => 1, mysql_enable_utf8 => 1 }
        );
    },
};

# Public

sub dbh {
    $dbh;
}

sub init ($) {
    $config = shift;
    die "Unsupported type of db" if ( !exists $dbi->{ $config->{db}->{type} } );
    $dbh = $dbi->{ $config->{db}->{type} }->( $config->{db} );
}

1;
__END__

=pod

=head1 NAME

Site::Engine::Database - simple support for SQL databases

=head1 DESCRIPTION

This module provide simple database support for MySQL and SQLite

    # MySQL
    my %config = (
        ...
        "db" => {
           "type"   => "mysql",
           "db"     => "db",
           "dbhost" => "mysql.server",  # optional
           "dbport" => 3306,            # optional
           "dbuser" => "dbuser",
           "dbpass" => "dbpass",
        },
        ...
    );

    # SQLite
    my %config = (
        ...
        "db" => {
           "type"   => "sqlite",
           "db_file"=> "/path/to/sqlite.db",
           "dbuser" => "",
           "dbpass" => "",
        },
        ...
    );

=head1 COPYRIGHT AND LICENSE

Copyright 2011 by crux E<lt>thecrux@gmail.comE<gt>

This module is free software and is published under the same terms as Perl itself.

=cut


lib/Site/Engine/Template.pm
package Site::Engine::Template;
use strict;
use warnings;
use Exporter qw( import );
our @EXPORT  = qw( escape_html clear_html );
our $VERSION = '0.01';

my $config;

# Public

sub init ($) {
    $config = shift;
}

sub escape_html ($) {
    my $data = shift;
    return "" if ( !defined $data );
    $data =~ s/&/&amp;/g;
    $data =~ s/</&lt;/g;
    $data =~ s/>/&gt;/g;
    $data =~ s/"/&quot;/g;
    $data =~ s/'/&#39;/g;
    return $data;
}

sub clear_html ($) {
    my $data = shift;
    return "" if ( !defined $data );
    $data =~ s/\r//g;
    $data =~ s/\<(br|p)\s?\/?\>/\n/g;
    $data =~ s/\<\/p\>/\n\n/g;
    $data =~ s/&\w+;/ /g;
    $data =~ s/\<.+?\>//sg;
    escape_html $data;
}

sub template {
    my ( $template, $vars, $conf ) = @_;
    my $data = &_template( $template, $vars );
    my $layout =
      ( exists $conf->{layout} )
      ? $conf->{layout}
      : $config->{layout} || "main";
    if ( defined $layout ) {
        $vars->{content} = $data;
        $data = &_template( $layout, $vars );
    }
    return $data;
}

# Private

sub _ifs ($$$$$$$) {
    my ( $not, $var, $subvar, $value, $subvalue, $block, $vars ) = @_;
    if ( !exists $vars->{$var} ) {
        return ($not) ? $block : "";
    }
    my $ret;
    if ( !defined($subvar) ) {
        $ret = $vars->{$var};
    }
    elsif ( ref $vars->{$var} eq "HASH" && exists $vars->{$var}->{$subvar} ) {
        $ret = $vars->{$var}->{$subvar};
    }
    elsif ( ref $vars->{$var} eq "ARRAY" && $subvar !~ /\D/ ) {
        $ret = $vars->{$var}->[$subvar];
    }
    if ( defined $ret ) {
        if ( defined $value && $value !~ /\D/ ) {
            $ret = ( $ret !~ /\D/ ) ? ( $ret == $value ) : 0;
        }
        elsif ( defined $value && exists $vars->{$value} ) {
            if ( defined $subvalue ) {
                $ret =
                    ( defined $vars->{$value}->{$subvalue} )
                  ? ( $ret == $vars->{$value}->{$subvalue} )
                  : 0;
            }
            else {
                $ret =
                  ( defined $vars->{$value} ) ? ( $ret == $vars->{$value} ) : 0;
            }
        }
        elsif ( defined $value ) {
            $ret = 0;
        }
    }
    $ret = !$ret if ($not);
    ($ret) ? $block : "";
}

sub _set ($$$) {
    my ( $var, $value, $vars ) = @_;
    $vars->{$var} = $value;
    "";
}

sub _inc ($$$$) {
    my ( $template, $var, $subvar, $vars ) = @_;
    if ( defined $var && exists $vars->{$var} ) {
        if ( !defined $subvar ) {
            $template = $vars->{$var};
        }
        elsif ( ref $vars->{$var} eq "HASH" && exists $vars->{$var}->{$subvar} )
        {
            $template = $vars->{$var}->{$subvar};
        }
        elsif ( ref $vars->{$var} eq "ARRAY" && $subvar !~ /\D/ ) {
            $template = $vars->{$var}->[$subvar];
        }
        else {
            return "";
        }
    }
    elsif ( !defined $template ) {
        return "";
    }
    &_template( $template, $vars );
}

sub _var ($$$$) {
    my ( $var, $subvar, $attr, $vars ) = @_;
    return "" if ( !exists $vars->{$var} );

    my $ret = "";
    if ( !defined($subvar) ) {
        $ret = $vars->{$var};
    }
    elsif ( ref $vars->{$var} eq "HASH" && exists $vars->{$var}->{$subvar} ) {
        $ret = $vars->{$var}->{$subvar};
    }
    elsif ( ref $vars->{$var} eq "ARRAY" && $subvar !~ /\D/ ) {
        $ret = $vars->{$var}->[$subvar];
    }
    if ( !defined $attr ) {
        $ret = escape_html $ret;
    }
    elsif ( $attr eq "raw" ) {
        $ret = $ret;
    }
    elsif ( $attr eq "clear" ) {
        $ret = clear_html $ret;
    }
    elsif ( $attr eq "clear_para" ) {
        $ret = clear_html $ret;
        $ret = substr $ret, 0, index( $ret, "\n\n" );
    }
    elsif ( $attr =~ /^length(\d+)/ ) {
        my $len = $1;
        $ret = substr( $ret, 0, $len ) . "..." if ( length $ret > $len );
        $ret = escape_html $ret;
    }
    else {
        $ret = escape_html $ret;
    }
    $ret;
}

{
    my $set_re = qr/\[%\sSET\s(\w+)\s?=\s?\"(.+?)\"\s%\]/;
    my $inc_re = qr/\[%\sINCLUDE\s(?:\"(\w+)\")?(?:(\w+)(?:\.(\w+))?)?\s%\]/;
    my $ifs_re =
qr/\[%\sIF\s(!\s?)?(\w+?)(?:\.(\w+?))?(?:==(\w+)(?:\.(\w+))?)?\s%\](.+?)\[%\sFI\s%\]/s;
    my $var_re = qr/\[%\s(\w+?)(?:\.(\w+))?(?:\s\|\s(\w+))?\s%\]/;
    my $oth_re = qr/\[%.+?%\]/;

    sub _vars ($$) {
        my ( $ref, $vars ) = @_;
        $$ref =~ s/$ifs_re/_ifs $1, $2, $3, $4, $5, $6, $vars/eg;
        $$ref =~ s/$set_re/_set $1, $2, $vars/eg;
        $$ref =~ s/$inc_re/_inc $1, $2, $3, $vars/eg;
        $$ref =~ s/$var_re/_var $1, $2, $3, $vars/eg;

        # remove all other tags
        $$ref =~ s/$oth_re//g;
    }
}

sub _foreach ($$$$) {
    my ( $hash_name, $array_name, $block, $vars ) = @_;
    my $output = "";
    return "" if ( !exists $vars->{$array_name} );
    foreach my $h_ref ( @{ $vars->{$array_name} } ) {
        $vars->{$hash_name} = $h_ref;
        my $copy = $block;
        _vars \$copy, $vars;
        $output .= $copy;
        delete $vars->{$hash_name};
    }
    return $output;
}

{
    my $for_re = qr/\[%\sFOREACH\s(\w+)\sIN\s(\w+)\s%\](.+?)\[%\sEND\s%\]/s;

    sub _template ($$) {
        my ( $template, $vars ) = @_;
        my $file = $config->{templates} . "/" . $template . ".tt";
        die "Template $template not found" if ( !-f $file );
        local $/;
        open my $fh, "<:encoding(UTF-8)", $file or die $!;
        my $data = <$fh>;
        close $fh;
        $data =~ s/$for_re/_foreach $1, $2, $3, $vars/eg;
        _vars \$data, $vars;
        return $data;
    }
}

1;
__END__

=pod

=head1 NAME

Site::Engine::Template - simple TT's like template engine

=head1 DESCRIPTION

This module provide simple support for templates.

=head1 METHODS

=head2 init( \%config )

init module with configuration

=head2 escape_html( $html )

escape html symbols

=head2 clear_html( $html )

try to convert html to plain text (remove tags and html entities)

=head2 template( $template, $vars, $conf )

create page from $template file with $vars and $conf

=over

=item $template - name of file (without extension .tt)

=item $vars - ref to hash with values

=item $conf - ref to hash with configuration ( for example layout )

=back

=head1 EXAMPLE

    template 'name_of_template_file', {
        data => "some data",
        array => [ "first" , "second" ],
        hash  => {
            this => "that",
        },
    }, { layout => 'name_of_template_of_layout' };

    Templates dir is fetched from config (key 'templates'), templates default extension is .tt

=head1 SYNTAX

=head2 [% var %] - substitute variable

=over

=item [% var %]      - SCALAR

=item [% var.4 %]    - ARRAY ( 4th element of array var )

=item [% var.prop %] - HASH  ( value of key 'prop' of hash var )

=back

=over

=item [% var %] - by default value is escaped ( by escape_html() )

=item [% var | raw %] - show variable as is ( no escaping )

=item [% var | clear %] - clear html (remove all tags and enities)

=item [% var | clear_para %] - clear html and return only first paragraph of text

=item [% var | lengthXX %] - return only first XX symbols (escaped)

=back

=head2 [% IF condition %] some data [% FI %] - condition for inserting data. Doesn't support nesting!!!

=over

=item [% IF var %] - check that var is true (not 0, not empty string, not undef)

=item [% IF !var %] - check that var is false

=item [% IF var==1 %] - check if condition var == number is true

=item [% IF var1==var2 %] - check if condition var1 == var2 is true

=back

=head2 [% SET var="string" %] - set var with value of "string"

=head2 [% INCLUDE var %] - include another template

=over

=item [% INCLUDE var %] - include template with name from variable var

=item [% INCLUDE "string" %] - include template with name "string"

=back

=head2 [% FOREACH element IN array %] ... [% END %] - foreach cycle.

    Doesn't support nesting!!!

=head1 BUGS and WORKAROUNDS

FOREACH and IF doesn't support nesting, but you can use INCLUDE


    Wrong !!!!                              Correct

    [%IF condition1 %]              |       [%IF condition1 %]
        data1                       |           data1
        [% IF confition2 %]         |           [% INCLUDE "template_with_confition2" %]
            data2                   |       [% FI %]
        [% FI %]                    |
    [% FI %]                        |       In file template_with_confition2.tt:
                                    |       [% IF confition2 %]
                                    |            data2
                                    |       [% FI %]

=head1 COPYRIGHT AND LICENSE

Copyright 2011 by crux E<lt>thecrux@gmail.comE<gt>

This module is free software and is published under the same terms as Perl itself.

=cut


lib/Site/Engine.pm
package Site::Engine;
use strict;
use warnings;
use Data::Dumper;
use Time::HiRes qw( gettimeofday tv_interval );
use CGI qw( path_info request_method );
use Encode;
use Site::Engine::Template;
use Site::Engine::Session;
use Site::Engine::Database;
use Exporter qw( import );
our @EXPORT =
  qw( header get post template start_site param session dump_env redirect config prefix database to_dumper layout escape url_escape upload );
our $VERSION = '0.01';

# Private

my (
    %headers,        $body,    @routes, $config, $path_info,
    $request_method, $session, $addr,   $ua,     $layout
);
my $prefix = "";

# Public
sub header ($;$) {
    my $header = shift;
    if ( defined $_[0] ) {
        $headers{$header} = $_[0];
    }
    else {
        return ( exists $headers{$header} )
          ? $headers{$header}
          : undef;
    }
}

sub param (;$) {
    Encode::decode( utf8 => CGI::param( shift() ) );
}

sub upload (;$) {
    CGI::upload( shift() );
}

sub escape ($) {
    escape_html( shift() );
}

sub url_escape ($) {
    CGI::escape( shift() );
}

sub session (;$$) {
    if ( scalar @_ == 1 && !defined $_[0] ) {
        Site::Engine::Session::destroy_session( $session, $addr );
        $session = undef;
    }
    else {
        my ( $id, $ret ) =
          Site::Engine::Session::session( $session, $addr, @_ );
        if ( !defined $session || $session ne $id ) {
            header "Set-Cookie" => CGI::cookie(
                -name  => "session",
                -value => $id,

                #-expires=>"+".$config->{session_expires}."s"
            );
        }
        return $ret;
    }
}

sub prefix ($) {
    $prefix = shift;
}

sub layout ($) {
    $layout = shift;
}

sub get ($$) {
    my ( $route, $sub ) = @_;
    push @routes, [ 'GET', $prefix, $route, $sub, $layout ];
}

sub post ($$) {
    my ( $route, $sub ) = @_;
    push @routes, [ 'POST', $prefix, $route, $sub, $layout ];
}

sub template ($$;$) {
    if ( !exists $_[1]->{prefix} ) {
        $_[1]->{prefix} = $prefix;
    }
    my $conf = ( scalar @_ == 3 ) ? pop(@_) : {};
    if ( !exists $conf->{layout} && defined $layout ) {
        $conf->{layout} = $layout;
    }
    Site::Engine::Template::template( @_, $conf );
}

sub dump_env {
    return join "", Dumper( \$config, \%ENV );
}

sub redirect ($) {
    my $location = shift;
    $location = $prefix . $location if ( $location !~ m'^\w+://' );
    header "Status"   => 303;
    header "Location" => $location;
    "";
}

sub config {
    $config;
}

sub database {
    dbh;
}

sub to_dumper {
    Dumper( shift() );
}

sub start_site ($) {
    $config         = shift;
    $path_info      = path_info();
    $request_method = request_method();
    $addr           = $ENV{"REMOTE_ADDR"};
    $ua             = $ENV{"HTTP_USER_AGENT"};
    $session        = CGI::cookie("session");
    %headers        = ();
    my $t0 = [gettimeofday];

    Site::Engine::Template::init($config);
    Site::Engine::Session::init($config);
    Site::Engine::Database::init($config);

    header "Content-Type" => "text/html";

    foreach my $route (@routes) {
        if (   $request_method eq $route->[0]
            && $path_info =~ /^$route->[1]$route->[2]$/ )
        {
            my @matches = ( $path_info =~ /^$route->[1]$route->[2]$/ );
            $prefix = $route->[1];
            $layout = $route->[4];
            eval { $body = $route->[3]->(@matches); };
            if ($@) {
                my $err = "";
                if ( $config->{debug} ) {
                    $err = escape_html($@);
                }
                $body = qq{
                    <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
                    <html><head>
                    <title>500 Internal Error</title>
                    </head><body>
                    <h1>Sorry, error</h1>
                    <pre>$err</pre>
                    <p>Please report to webmaster.</p>
                    </body></html>
                };
                $body =~ s/\n\s+/\n/sg;
                header "Status" => 500;
            }
            last;
        }
    }
    if ( !defined $body  ) {
        $body = qq{
            <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
            <html><head>
            <title>404 Not Found</title>
            </head><body>
            <h1>Not Found</h1>
            <p>The requested URL was not found on this server.</p>
            </body></html>
        };
        $body =~ s/\n\s+/\n/sg;
        header "Status" => 404;
    }
    header "X-Elapsed-Time" => tv_interval($t0);

    binmode STDOUT, ":utf8";

    print join "\n", map { $_ . ": " . $headers{$_} } keys %headers;
    print "\n\n";
    print $body;
}

1;
__END__

=pod

=head1 NAME

Site::Engine - Ugly tiny CGI-based web-framework with templates, sessions and databases.

=head1 DESCRIPTION

This is simple CGI-based web-framework. Use this only if your hosting has plain perl and
Apache with mod_cgi/mod_rewrite and doesn't provide support for modern PSGI-based WF
like Dancer or Mojolicious

=head1 SYNOPSIS

    #!/usr/bin/perl
    use strict;
    use warnings;
    use Site::Engine;

    # Configuration of site
    my %config = (
        "templates" => "/path/to/templates/dir";
        "db" => {
           "type"   => "mysql",
           "db"     => "db",
           "dbuser" => "dbuser",
           "dbpass" => "dbpass",
        },
        "debug" => 1
    );

    # Routes
    get qr{/} => sub {
        template 'index', {
            data => "hello world!"
        }, { layout => 'main' }
    };

    post qr{/login} => sub {
        my $user = param('user');
        my $pass = paran('pass');
        if ($user eq "admin" && $pass eq "secret") {
            session "user" => "admin";
            redirect '/';
        } else {
            redirect '/?error=badpass'
        }
    };

    get qr{/(\w+)} {
        my $match = shift;
    };

    # Start engine

    start_site \%config;

=head1 METHODS

=head2 start_site \%config

start processing request

=head2 header $header [, $value]

get/set header

=head2 param $param

alias of CGI::param (return utf-8 encoded string)

=head2 upload $file

alias of CGI::upload (return byte string)

=head2 escape $string

escape html symbols in string

=head2 url_escape $url

alias of CGI::escape()

=head2 session [ $key [, $value] ]

get/set session key

To destroy session:

    session undef;

=head2 prefix $prefix

Set prefix for routes

    prefix "/admin";

    get qr{/login} => sub { # Match /admin/login
    }

=head2 layout $layout

Set default layout for routes

=head2 get

Defines a route for HTTP GET requests to the given path:

    get qr{/} => sub {
        "Hello, world!"
    };

    All matches in regexp will be availiable via @_

    get qr{/name/(\w+)} => sub {
        my $name = shift;
    };

=head2 post

Defines a route for HTTP POST requests to the given path:

    post qr{/login} => sub {
        my $login = param("login");
        my $pass  = param("pass");
    }

=head2 template

Build page with specified template

    template 'name_of_template_file', {
        var => "value",
        array => [ "some", "array" ],
        hash => { "this" => "that" },
    }, { layout => 'layout_template' };

=head2 dump_env

return Data::Dumper for $config and \%ENV

=head2 redirect

return 303 redirect to URL (if prefix is set, than $prefix+$URL)

=head2 config

return ref to config hash

=head2 database

return database handle (dbh)

=head2 to_dumper

alias of Data::Dumper()

=head1 COPYRIGHT AND LICENSE

Copyright 2011 by crux E<lt>thecrux@gmail.comE<gt>

This module is free software and is published under the same terms as Perl itself.

=cut
